firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java

```diff
@@ -2247,4 +2274,205 @@ public void deserializationFailureIncludesPath() {
           e.getMessage());
     }
   }
+
+  // Bean definitions with @DocumentId applied to wrong type.
+  private static class FieldWithDocumentIdOnWrongTypeBean {
+    @DocumentId public Integer intField;
+  }
+
+  private static class SetterWithDocumentIdOnWrongTypeBean {
+    private int intField = 100;
+
+    @DocumentId
+    public void setIntField(int value) {
+      intField = value;
+    }
+  }
+
+  private static class GetterWithDocumentIdOnWrongTypeBean {
+    private int intField = 100;
+
+    @DocumentId
+    public int getIntField() {
+      return intField;
+    }
+  }
+
+  private static class PropertyWithDocumentIdOnWrongTypeBean {
+    @PropertyName("intField")
+    @DocumentId
+    public int intField = 100;
+  }
+
+  @Test
+  public void documentIdAnnotateWrongTypeThrows() {
+    assertExceptionContains(
+        "instead of String or DocumentReference",
+        () -> serialize(new FieldWithDocumentIdOnWrongTypeBean()));
+    assertExceptionContains(
+        "instead of String or DocumentReference",
+        () -> deserialize("{'intField': 1}", FieldWithDocumentIdOnWrongTypeBean.class));
+
+    assertExceptionContains(
+        "instead of String or DocumentReference",
+        () -> serialize(new SetterWithDocumentIdOnWrongTypeBean()));
+    assertExceptionContains(
+        "instead of String or DocumentReference",
+        () -> deserialize("{'intField': 1}", SetterWithDocumentIdOnWrongTypeBean.class));
+
+    assertExceptionContains(
+        "instead of String or DocumentReference",
+        () -> serialize(new GetterWithDocumentIdOnWrongTypeBean()));
+    assertExceptionContains(
+        "instead of String or DocumentReference",
+        () -> deserialize("{'intField': 1}", GetterWithDocumentIdOnWrongTypeBean.class));
+
+    assertExceptionContains(
+        "instead of String or DocumentReference",
+        () -> serialize(new PropertyWithDocumentIdOnWrongTypeBean()));
+    assertExceptionContains(
+        "instead of String or DocumentReference",
+        () -> deserialize("{'intField': 1}", PropertyWithDocumentIdOnWrongTypeBean.class));
```

Comment:
nit: Test boilerplate (repetition) is one of my big pet peeves.  It makes it harder to change implementation details because there are lots of tests to update for a single implementation change.  We end up with lots of repetition anyway, but I think it's always worth asking if/how it can be avoided.  So consider:
1) Putting the expected error text into a variable that's re-used, or even less repetition:
2) Create a list of lambdas containing all of these operations and just loop through them, rather than call assertExceptionContains() a bunch of times.

Note that 2 could make the test a bit harder to debug if it fails, so it's a trade-off.  You could also create a helper method (or local lambda function) that wraps the `assertExceptionContains("instead of String or DocumentReference", ...)`  call and just call it repeatedly.

Reply:
Implemented #1.

I think `assertExceptionContains` here is pretty short and straightforward, adding an extra wrapping leads to more cognition load without a lot benefits. And it's harder to debug too, as you pointed out.